### PR TITLE
cmd/containerboot: optionally serve health check endpoint

### PR DESCRIPTION
**Pull request by irbekrm**
Add functionality to optionally serve an HTTP health check endpoint (off by default) for containerized tailscale.

Users can enable health check endpoint by setting TS_HEALTHCHECK_ADDR_PORT env var to `[<addr>]:<port>`.

Containerboot will then serve an unauthenticatd HTTP health check at `/healthz` at that address.
The health check returns 200 OK if the node has at least one tailnet IP address, else returns 503.

I have tested that this works as a backend target HTTP health check for AWS NLB.

Also have tested it as a [readiness probe for Kubernetes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes):
```
...
containers:
  - name: tailscale
    env:
    - name: POD_IP
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: status.podIP
    - name: TS_HEALTHCHECK_ADDR_PORT
      value: $(POD_IP):9002
    readinessProbe:
      httpGet:
        path: /healthz
        port: 9002
...
```

Updates tailscale/tailscale#12898

### CODE REVIEW

File: cmd/containerboot/main.go [hunk 1]
```diff
@@ -1374,3 +1398,41 @@ func tailscaledConfigFilePath() string {
...
+	hasAddrs bool
+}
+
+func (h *healthz) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.Lock()
+	defer h.Unlock()
+	if h.hasAddrs {
+		w.Write([]byte("ok"))
+	} else {
+		http.Error(w, "node currently has no tailscale IPs", http.StatusInternalServerError)
```
Comment (irbekrm) [inline cmd/containerboot/main.go:1416]: Not sure if there is a more suited status code than 503

File: cmd/containerboot/main.go [hunk 2]
```diff
@@ -565,6 +576,13 @@ runLoop:
 						log.Fatalf("storing device IPs and FQDN in Kubernetes Secret: %v", err)
 					}
 				}
+
+				if cfg.HealthCheckAddrPort != "" {
+					h.Lock()
+					h.hasAddrs = len(addrs) != 0
```
Comment (irbekrm) [inline cmd/containerboot/main.go:582]: Alternatively, we could consider the node ready when it has endpoints. I am not sure at the moment what would be the best, so used tailnet addresses- we could change it in the future when we get some feedback

File: cmd/containerboot/main.go [hunk 3]
```diff
@@ -52,6 +52,12 @@
 //     ${TS_CERT_DOMAIN}, it will be replaced with the value of the available FQDN.
 //     It cannot be used in conjunction with TS_DEST_IP. The file is watched for changes,
 //     and will be re-applied when it changes.
+//   - TS_HEALTHCHECK_ADDR_PORT: if specified, an HTTP health endpoint will be
+//     served at /healthz at the provided address, which should be in form [<address>]:<port>.
+//     If not set, no health check will be run. If set to :<port>, addr will default to 0.0.0.0
```
Comment (jaxxstorm) [inline cmd/containerboot/main.go:57]: do we need to specify the addr here in the comment?

**Review by jaxxstorm — APPROVED** [review]

File: cmd/containerboot/main.go (hunk 3 above)
Comment (irbekrm) [reply cmd/containerboot/main.go:57]: > do we need to specify the addr here in the comment?

Could you clarify- do you mean the comment should have less info or more?

Comment (irbekrm) [reply cmd/containerboot/main.go:57]: I'm going to merge this now, but we can improve the comment.

**Pull request merged**